pallets/avn/src/lib.rs

```diff
@@ -246,28 +273,62 @@ impl<T: Config> Pallet<T> {
 
     // TODO [TYPE: refactoring][PRI: LOW]: choose a better function name
     pub fn is_primary(
-        block_number: BlockNumberFor<T>,
+        op_type: OperationType,
         current_validator: &T::AccountId,
     ) -> Result<bool, Error<T>> {
-        let primary_validator = Self::calculate_primary_validator(block_number)?;
-        return Ok(&primary_validator == current_validator)
+        let validators = Self::validators();
+
+        // If there are no validators there's no point continuing
+        if validators.len() == 0 {
+            return Err(Error::<T>::NoValidatorsFound)
+        }
+
+        let counters = PrimaryCollator::<T>::get();
+
+        let index = match op_type {
+            OperationType::Ethereum => counters.ethereum as usize,
+            OperationType::Avn => counters.avn as usize,
+        };
+
+        if index >= validators.len() {
+            return Err(Error::<T>::InvalidValidatorIndex)
+        };
+
+        let primary_validator = &validators[index].account_id.clone();
+
+        return Ok(primary_validator == current_validator)
     }
 
-    pub fn calculate_primary_validator(
-        block_number: BlockNumberFor<T>,
-    ) -> Result<T::AccountId, Error<T>> {
+    pub fn advance_primary_validator(op_type: OperationType) -> Result<T::AccountId, Error<T>> {
         let validators = Self::validators();
 
         // If there are no validators there's no point continuing
         if validators.len() == 0 {
             return Err(Error::<T>::NoValidatorsFound)
         }
 
-        let block_number: usize = TryInto::<usize>::try_into(block_number)
-            .map_err(|_| Error::<T>::ErrorConvertingBlockNumber)?;
-
-        let index = block_number % validators.len();
-        return Ok(validators[index].account_id.clone())
+        let mut counters = PrimaryCollator::<T>::get();
+        let validators_len = Self::validators().len() as u8;
+
+        let index = match op_type {
+            OperationType::Ethereum => {
+                counters.ethereum = (counters.ethereum + 1) % validators_len;
+                PrimaryCollator::<T>::put(PrimaryCollatorData {
+                    ethereum: counters.ethereum,
+                    avn: counters.avn,
+                });
+                counters.ethereum
+            },
+            OperationType::Avn => {
+                counters.avn = (counters.avn + 1) % validators_len;
```

Comment:
can use saturating here as well.

Reply:
We should not need to obsess about saturations, should we? Why this push for using it everywhere?
The `%` ensures the numbers will never be larger than `validators_len`. Realistically, for us that is as small as it gets (`u8`). Why do we need to obsess about saturation at all?